### PR TITLE
[MTP] Force greedy sampling on AMD

### PR DESCRIPTION
**Pull request by datdo-msft**
## Motivation

In EAGLE utils, CUDA kernels are used as part of sampling tokens, specifically getting target probs here: https://github.com/valarLip/sglang/blob/8d463fe351c4b10ed8b65d7fe5ed0c68eea6d731/python/sglang/srt/speculative/eagle_utils.py#L370. There are already [existing checks](https://github.com/valarLip/sglang/blob/8d463fe351c4b10ed8b65d7fe5ed0c68eea6d731/python/sglang/srt/speculative/eagle_utils.py#L24) for `is_cuda()` and `is_hip()` which dictates which kernels get imported. As a result, we should treat all requests as greedy sampling on AMD.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

**Comment by datdo-msft** [comment]
Hi @Ying1123 and @merrymercy , please help review and let me know if this makes sense. Thanks!

File: python/sglang/srt/speculative/eagle_utils.py [hunk 1]
```diff
@@ -410,8 +410,8 @@ def verify(
                 logits=logits_output.next_token_logits, vocab_mask=vocab_mask
             )
 
-        # Sample tokens
-        if batch.sampling_info.is_all_greedy:
+        # Sample tokens. Force greedy sampling on AMD
+        if batch.sampling_info.is_all_greedy or is_hip():
```
[the remark this answers is not in the document]
Comment (HaiShaw) [reply python/sglang/srt/speculative/eagle_utils.py]: @datdo-msft please take this suggest change, so that when forward improvement is made, we will have it more clearly.

**Review by HaiShaw — CHANGES_REQUESTED** [review]
It is w.i.p. Still better to take code-assist's

**Comment by datdo-msft** [comment]
@HaiShaw , updated please help take a look again. Thanks!

**Review by HaiShaw — APPROVED** [review]
LGTM

**Comment by hubertlu-tw** [comment]
@datdo-msft could you please run the following commands to fix lint errors:
```
pip3 install pre-commit
pre-commit run --all-files
```

File: python/sglang/srt/speculative/eagle_utils.py [hunk 2]
```diff
@@ -410,8 +410,11 @@ def verify(
                 logits=logits_output.next_token_logits, vocab_mask=vocab_mask
             )
 
-        # Sample tokens
-        if batch.sampling_info.is_all_greedy:
+        # Sample tokens. Force greedy sampling on AMD
```
Comment (merrymercy) [inline python/sglang/srt/speculative/eagle_utils.py:428]: We should print a warning here

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: python/sglang/srt/speculative/eagle_utils.py (hunk 2 above)
Comment (datdo-msft) [reply python/sglang/srt/speculative/eagle_utils.py:428]: Done

File: python/sglang/srt/speculative/eagle_utils.py [hunk 3]
```diff
@@ -423,8 +423,15 @@ def verify(
...
+        if (
+            batch.sampling_info.is_all_greedy
+            or "tree_speculative_sampling_target_only" not in globals()
+        ):
+            logger.warning(
```
Comment (merrymercy) [inline python/sglang/srt/speculative/eagle_utils.py:431]: Only print this warning for amd!

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: python/sglang/srt/speculative/eagle_utils.py [hunk 4]
```diff
@@ -423,8 +425,15 @@ def verify(
                 logits=logits_output.next_token_logits, vocab_mask=vocab_mask
             )
 
-        # Sample tokens
-        if batch.sampling_info.is_all_greedy:
+        # Sample tokens. Force greedy sampling on AMD
+        is_all_greedy = sampling_info.is_all_greedy
+        if (not is_all_greedy) and (not TREE_SPEC_KERNEL_AVAILABLE) and is_hip():
```
Comment (merrymercy) [inline python/sglang/srt/speculative/eagle_utils.py]: drop `is_hip`

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: python/sglang/srt/speculative/eagle_utils.py (hunk 4 above)
Comment (datdo-msft) [reply python/sglang/srt/speculative/eagle_utils.py]: Done

**Review by merrymercy — APPROVED** [review]
